### ISSUE TITLE
Server sends Handshake probe packet which does not attribute to any progress

### ISSUE THREAD

**Issue by tatsuhiro-t** [quicwg/base-drafts#3582]
Suppose the following situation:

- Upon reception of client Initial, server sent Handshake(server_handshake#0) and Short(server_short#0) packets.
- server_short#0 was lost.
- client received server_handshake#0 and sent packet containing ClientFinished Handshake(client_handshake#0)
- client_handshake#0 was lost.
- server resent Handshake(server_handshake#1)
- client received server_handshake#1 and sent packet containing ACK only Handshake(client_handshake#1)
- server received client_handshake#1

At this point, server has still bytes_in_flight > 0 because it
has in-flight Short packet.  According to the current recovery
draft, OnLossDetectionTimeout instructs server to send Handshake
probe packet because handshake has not completed yet.  I'm not
sure what this packet is intended for. server has no in-flight
data for Handshake packet number space.  Client will send ACK for
it, but it does not progress any handshake state.

**Comment by martinthomson** [comment]
Presumably the server will be probing for server_short[0] here.  The handshake probe is recommended, but I agree that if the handshake is done, there is no need for it, but it isn't mandatory.

**Comment by ianswett** [comment]
Q: I assume the server sent and the client received server_initial#0 as well in this example?

The interesting text is:
> When ack-eliciting packets are in-flight in multiple packet number spaces, the timer MUST be set for the packet number space with the earliest timeout, except for ApplicationData, which MUST be ignored until the handshake completes; see Section 4.1.1 of {{QUIC-TLS}}. Not arming the PTO for ApplicationData prevents a client from retransmitting a 0-RTT packet on a PTO expiration before confirming that the server is able to decrypt 0-RTT packets, and prevents a server from sending a 1-RTT packet on a PTO expiration before it has the keys to process an acknowledgement.

In this case, the text says to not arm PTO for ApplicationData until handshake complete and there's nothing in flight in Initial or Handshake, so I don't believe the PTO alarms should be set, so "server resent Handshake(server_handshake#1)" would never occur.

Admittedly, this text can be hard to read(it's certainly hard to write), so if you have any editorial suggestions, that'd be welcome.

**Comment by tatsuhiro-t** [comment]
The text looks good.  It looks like the pseudo code does not capture it though.  Currently, it says:

```py
  if (no ack-eliciting packets in flight &&
      PeerCompletedAddressValidation()):
```

There is no condition to check handshake completion there.  I'm not sure how the pseudo code should be precise in this regard.

**Comment by ianswett** [comment]
Good point, I tried to fix that in #3604
